doc/man/doodle.1
.TH DOODLE "1" "Jan 1 2010" "doodle"

.SH "NAME"
doodle \- a tool to search the meta\-data in your files

.SH "SYNOPSIS"
.B doodle
[\fIOPTIONS\fR] ([FILENAMES]*|[KEYWORDS]*)

.SH "DESCRIPTION"
.PP
doodle is a tool to index files.  doodle uses libextractor to find meta\-data in files.  Once a database has been built, doodle can be used to quickly find files of which the meta-data matches a given search\-string.  This way, doodle can be used to quickly search your file system.
.TP
Generally, the first time you run doodle you pass the option \-b to build the database.  Together with \-b you specify the list of files or directories to index, for example 

$ doodle \-b $HOME

.TP
Indexing with doodle is incremental.  If doodle \-b is run (with the same database) twice it will update the index for files that were changed.  doodle will also remove files that are no longer accessible.  doodle will NOT remove files that are still present but no longer specified in the argument list.  Thus invoking either

$ doodle \-b /foo /bar  # or

$ doodle \-b /foo ; doodle \-b /bar

.TP
will result in the same database containing both the index for /foo and /bar.  Note that the only way to only un-index /foo at this point is to make /foo inaccessible (using for example chmod 000 /foo or even rm -rf /foo) and then run doodle \-b again.

.TP
In networked environments, it often makes sense to build a database at the root of each file system, containing the entries for that  file system.  For this, doodle is run for each file system on the file server where that file system is on a local disk, to prevent thrashing the network.  Users can select which databases doodle searches. Databases cannot be concatenated together.

.TP
Once the files have been indexed, you can quickly query the doodle database.  Just run

$ doodle keyword

.TP
to search all of your files for keyword.  Note that only the meta-data extracted by libextractor is searched.  Thus if libextractor does not find any meta-data in the files, you may not get any results.  You can use the option \-l to specify non-standard libextractor plugins.  For example, doodle could be used to replace the locate tool from the GNU findutils like this:

$ alias updatedb="doodle \-bn \-d ~/.doodle\-locate\-db \-l libextractor_filename /"

$ alias locate="doodle \-d ~/.doodle\-locate\-db"


.SH "OPTIONS"
.TP
\fB\-a \fINUMBER\fR, \fB\-\-approximate=\fINUMBER\fR
do approximate matching with mismatches of up to NUMBER letters
.TP
\fB\-b, \fB\-\-build\fR
build the doodle database (passed arguments are directories and filenames that are to be indexed).  In comparison with GNU locate the doodle binary encapsulates both the locate and the updatedb tool.  Using the \fB\-b\fR option doodle builds or updates the database (equivalent to updatedb), without \fB\-b\fR it behaves similar to locate.
.TP 
\fB\-d \fIFILENAME\fR, \fB\-\-database=\fIFILENAME\fR
use FILENAME for the location of the database (use when building or searching).  This option is particularly useful when doodle is used to search different types of files (or is operated with different extractor options).  Using this option doodle can be used to build specialized indices (i.e. one per file system), which can in turn improve search performance.  When searching, you can pass a colon-separated list of database file names, in that case all databases are searched.  Note that the disk-space consumption of a single database is typically slightly smaller than if the database is split into multiple files.  Nevertheless, the space\-savings are likely to be small (a few percent).  You can also use  the environment variable DOODLE_PATH to set the list of database files to search.  The option overrides the environment variable if both are used.  If the option is not given and DOODLE_PATH is not set, "~/.doodle" is used.
.TP 
\fB\-e\fR, \fB\-\-extract\fR
print the extracted keywords for each matching file found.  Note that this will slow down the program a lot, especially if there are many matches in the database.  Note that if the options given for libextractor are different than the options used for building the index the results may not contain the search string.
.TP
\fB\-f\fR, \fB\-\-filenames\fR
include filenames (full path) in the set of keywords
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-i, \fB\-\-ignore\-case\fR
be case-insensitive
.TP
\fB\-l \fILIBRARIES\fR, \fB\-\-library=\fILIBRARIES\fR
specify which libextractor plugins to use (for building the index with \-b or for printing information about files with \-e)
.TP
\fB\-L \fIFILENAME\fR, \fB\-\-log=\fIFILENAME\fR
log all encountered keywords into a log file named FILENAME.  This option is mostly useful for debugging.
.TP
\fB\-m \fILIMIT\fR\fR, \fB\-\-memory=\fILIMIT\fR
use at most LIMIT MB of memory for the nodes of the suffix-tree (after that, serialize to disk).  Note that a smaller value will reduce memory consumption but increase the size of the temporary file (and slow down indexing).  The default is 8 MB.
.TP
\fB\-n\fR, \fB\-\-nodefault\fR
do not load the default set of plugins (only load plugins specified with \-l)
.TP
\fB\-p\fR, \fB\-\-print\fR
make a human-readable screen dump of the doodle database (only really useful for debugging)
.TP
\fB\-P \fIPATH\fR, \fB\-\-prunepaths=\fIPATH\fR
Directories to not put in the database, which would otherwise be. The environment variable PRUNEPATHS also sets this value. Default is "/tmp /usr/tmp /var/tmp /dev /proc /sys".  This option can also be used when searching, in which case search results in the specified directories will be ignored.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-V\fR, \fB\-\-verbose\fR
be verbose

.SH "ENVIRONMENT"
.TP
.B DOODLE_PATH
Colon\-separated list of databases to search.  Note that when building the database this path must either only contain one filename or the option \fB\-b\fP must be used to specify the database file.  Default is "~/.doodle".
.TP
.B PRUNEPATHS
Space\-separated list of paths to exclude.  Can be overridden with the \fB\-P\fR option.  

.SH "NOTES"
Doodle depends on libextractor.  You can download libextractor
from http://gnunet.org/libextractor/.

.SH "SEE ALSO"
\fBextract\fP(1), \fBslocate\fP(1), \fBupdatedb\fP(1), \fBlibextractor\fP(3), \fBlibdoodle\fP(3)

.SH "LEGAL NOTICE"
libdoodle and doodle are released under the GPL.

.SH "REPORTING BUGS"
Report bugs to mantis <https://gnunet.org/bugs/> or by sending electronic mail to <christian@grothoff.org>

.SH "AUTHORS"
doodle was originally written by Christian Grothoff <christian@grothoff.org>.

.SH "AVAILABILITY"
You can obtain the original author's latest version from http://grothoff.org/christian/doodle/.


ChangeLog
Thu Jan 14 11:53:06 CET 2010
	Releasing doodle 0.7.0.
	
Fri Jan  1 21:07:59 CET 2010
	Updating implementation for GNU libextractor 0.6.x API.

Mon Apr 28 08:41:16 MDT 2008
	Updated Dutch translation.

Sun Apr 20 20:52:16 MDT 2008
	Fixed a problem in the code handling iconv (character
	set translation) errors; if iconv was not working, the
	input string should just be returned without conversion
	(thanks to Alan Post for pointing this out).

Wed Dec 19 11:06:08 MST 2007
	Updated Vietnamese translation.

Mon Dec 10 17:10:08 MST 2007
	Updated Swedish translation.

Wed Dec  5 18:26:24 MST 2007
	Added "missing" include file for pthread_kill.
	Releasing doodle 0.6.7.

Tue Nov  6 12:01:10 MST 2007
	Updated Dutch translation.

Mon Feb 12 13:44:03 MST 2007
	gettext update to 0.16.1.

Tue Jan  2 21:25:13 MST 2007
	Patching bug on big-endian platforms (thanks to Volker Weiss).
	This changes the database format (for everybody), so you will
	need to regenerate your indices.
	Patching bug in handling of empty metadata (again, thanks to
	Volker Weiss).
	Releasing doodle 0.6.6.
	
Mon Jan  1 19:10:40 MST 2007
	Added pkgconfig support.

Tue May  2 14:56:42 PDT 2006
	doodled shutdown sequence was broken, fixed.
	doodled used stat instead of lstat, resulting in
	symbolic links being added and then removed from
	the database; fixed.  doodled no longer verifies
	all entries in the database on startup (too costly).
	We assume that "doodle -b" has been used to create
	an up-to-date database and that doodled will always
	run in the background keeping it (sufficiently)
	up-to-date.
	Added out-of-process execution of GNU libextractor.
	This ensures that doodle(d) does not crash even if
	GNU libextractor does (i.e. due to bugs or corrupted
	file systems).
	Releasing doodle 0.6.5.

Sat Jan  7 02:37:53 PST 2006
	Added Swedish translation.

Sun Sep 18 22:14:17 PDT 2005
	Actually implemented keyword logging.  Releasing doodle 0.6.4.

Sun Sep 18 16:30:56 PDT 2005
	Updated Irish translation.

Fri Sep 16 19:06:42 PDT 2005
	Updated German translation.

Thu Sep 15 21:55:19 PDT 2005
	Fixing compiler warnings given by gcc 4.0.

Sun Sep  4 17:56:10 PDT 2005
	Fixed -P option (short form did not work, Mantis #927).
	Made prune-paths much faster by not iterating over the pruned
	directories (Mantis #926).

Wed Aug  3 11:01:18 PDT 2005
	Updated Vietnamese translation.

Thu Jul 28 16:35:50 EST 2005
	Small changes to support API of libextractor >= 0.5.0.
	Releasing doodle 0.6.3.

Tue Apr  5 17:11:33 EST 2005
	Added Vietnamese translation.

Mon Feb  7 11:06:05 EST 2005
	Added /sys to default prunepath.

Wed Jan 26 03:20:11 EST 2005
	Added Italian translation.

Wed Jan  5 17:34:06 EST 2005
	Fixed bug with approximate searches trying to write to the database.
	Releasing doodle 0.6.2.  (SVN revision is 52).
	
Fri Dec 31 20:33:44 EST 2004
	Updated Irish and German translation.

Mon Dec 27 21:21:35 CET 2004 
	Do not follow symbolic links (in doodled, doodle was ok already).

Sun Dec 26 01:44:00 CET 2004
	doodled debugging.  Added multithreading, code cleanup.  Works much
	better now.  Added conversion of search strings to UTF-8 to support
	libextractor 0.4.0 and higher.  Releasing doodle 0.6.1.

Sat Nov 13 15:00:03 EST 2004
	Added locking (flock) to avoid concurrent modification of the doodle
	DB.  More doodled bugfixes and optimizations.  In particular fine-grained
	modification tracking (per node) and batched truncation.  This
	dramatically speeds up updating an existing database after files were
	removed (O(n) instead of O(n*k) where n is the size of the DB and k
	is the number of removed files).

Thu Nov 11 21:50:01 EST 2004
	Various major bugfixes to doodled and also in doodle
	(with respect to logging).

Fri Oct 22 22:04:32 EST 2004
	Added doodled, a daemon that watches changes in directories
	that were indexed by doodle and updates the database whenever
	files in those directories change.  Highly experimental code.
	Uses libfam.  Releasing doodle 0.6.0.
	
Thu Oct 21 21:55:31 EST 2004
	Major bugfixes in the new code.  The format was again
	changed slightly.  Where in 0.5.0 all offsets were relative,
	this is no longer the case for the special absolute offset "0"
	which is used for termination. Since "0" is not a valid relative
	offset (a node may not point to itself) we do not have to encode
	0 as a relative offset and can just put an absolute 0.  This
	safes a couple of bytes.  Also we did not check if the same
	node was listed (by chance) multiple times for the same filename.
	Again, doing that check costs hardly any time but improves 
	the database size.  For a small testcase we're now down
	20% in DB size from 0.5.0 (270458 bytes vs. 223311 bytes).

Wed Oct 20 21:33:34 EST 2004
	Added multi-link support.  Multi-link support allows
	much faster scanning of the doodle tree by allocating
	n-nodes from the same level of the tree in the same
	block in memory.  This changes the database format
	slightly; the cost and benefits in terms of size should
	roughly cancel each other out (on some test I saw an
	increase of 1.9%).  For searching this should not make 
	much of a difference; single search invocations
	should be minimally slower, on the other hand multiple 
	searches on the same tree in memory should be faster.
	In general, building the database should be significantly
	faster (at least as far as doodle is concerned) since in 
	essence this entails doing lots of searches in-memory.
	Using the proftree benchmark the time for 100.000 
	in-memory 2-searches goes down from 150s to 110s on my 
	machine.

	Did I mention that this feature is highly experimental and
	the current CVS code is thus far from stable (for example
	proftree2 still crashes!) If you want something stable, 
	stick to doodle 0.5.0 for now!

Sat Oct 16 15:32:48 EST 2004
	Added translation to Albanian.  
	
Sat Oct  9 15:31:18 EST 2004
	Added translation to Dutch.  Releasing 0.5.0.

Tue Oct  5 02:05:02 EST 2004

src/doodle/doodle.c
/*
     This file is part of doodle.
     (C) 2004, 2010 Christian Grothoff (and other contributing authors)

     doodle is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     doodle is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with doodle; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file doodle/doodle.c
 * @brief main method of doodle (binary).
 * @author Christian Grothoff
 *
 * doodle links against libdoodle to use the code factored into
 * tree.c.  doodle uses GNU getopt and for portability ships with its
 * own version of the getopt code (getopt*).<p>
 *
 * doodle requires libextractor (http://gnunet.org/libextractor/).
 */

#include "config.h"
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <fcntl.h>
#include <errno.h>
#include <netinet/in.h>
#include <stdarg.h>
#include <dirent.h>
#include <extractor.h>
#include <locale.h>
#include <langinfo.h>

#include "doodle.h"
#include "gettext.h"
#include "getopt.h"
#include "helper2.h"
#include "convert.h"


static int verbose = 0;
static int very_verbose = 0;

/**
 * Print the doodle-specific text for --help.
 */
static void printHelp () {
  static Help help[] = {
    { 'a', "approximate", "DISTANCE",
      gettext_noop("consider strings to match if DISTANCE letters are different") },
    { 'b', "build", NULL,
      gettext_noop("build database (default is to search)") },
    { 'd', "database", "FILENAME",
      gettext_noop("use location FILENAME to store doodle database") },
    { 'e', "extract", NULL,
      gettext_noop("for each matching file, print the extracted keywords") },
    { 'f', "filenames", NULL,
      gettext_noop("add the filename to the list of keywords (use when building database)") },
    { 'h', "help", NULL,
      gettext_noop("print this help page") },
    { 'i', "ignore-case", NULL,
      gettext_noop("be case-insensitive (use when searching)") },
    { 'l', "library", "LIBRARY",
      gettext_noop("load an extractor plugin named LIBRARY") },
    { 'L', "log", "FILENAME",
      gettext_noop("log keywords to a file named FILENAME") },
    { 'n', "nodefault", NULL,
      gettext_noop("do not load default set of extractor plugins") },
    { 'm', "memory", "SIZE",
      gettext_noop("set the memory limit to SIZE MB (for the tree).") },
    { 'p', "print", NULL,
      gettext_noop("print suffix tree (for debugging)") },
    { 'P', "prunepaths", NULL,
      gettext_noop("exclude given paths from building or searching") },
    { 'v', "version", NULL,
      gettext_noop("print the version number") },
    { 'V', "verbose", NULL,
      gettext_noop("be verbose") },
    { 0, NULL, NULL, NULL },
  };
  formatHelp(_("doodle [OPTIONS] ([FILENAMES]*|[KEYWORDS]*)"),
	     _("Index and search meta-data of files."),
	     help);
}

static int do_extract = 0;
static int do_default = 1;
static int do_print   = 0;
static int do_filenames = 0;
static int ignore_case = 0;
static unsigned int do_approx = 0;
static char * prunepaths = "/tmp /usr/tmp /var/tmp /dev /proc /sys";

/* *************** helper functions **************** */

/**
 * Print log-messages to stdout.
 */
static void my_log(void * unused,
		   unsigned int level,
		   const char * msg,
		   ...) {
  va_list args;
  if ( (level == 0) ||
       (verbose && (level == 1)) ||
       (very_verbose && (level == 2) ) ) {
    va_start(args, msg);
    vfprintf(stdout, msg, args);
    va_end(args);
  }
}

static int isPruned(const char * filename,
		    void * cls) {
  int i;
  int last;

  i = strlen(prunepaths);
  last = i;
  while (i > 0) {
    while ( (prunepaths[i] != ' ') && (i >= 0) )
      i--;
    if (0 == strncmp(&prunepaths[i+1],
		     filename,
		     last - (i+1))) {
      if (very_verbose)
	printf(_("Pruned: %s\n"), filename);
      return 1;
    }
    last = i;
    i--;
  }
  return 0;
}


/* ************** actual functionality code ************* */


/**
 * @brief closure for do_index
 */
typedef struct {
  struct EXTRACT_Process * elist;
  struct DOODLE_SuffixTree * tree;
  FILE * logFile;
} DIC;


static int do_index(const char * filename,
		    void * cls) {
  DIC * dic = cls;
  int i;
  int j;
  struct stat sbuf;

  if (isPruned(filename, NULL))
    return 0;
  j = -1;
  for (i=DOODLE_getFileCount(dic->tree)-1;i>=0;i--)
    if (0 == strcmp(filename,
		    DOODLE_getFileAt(dic->tree,i)->filename)) {
      j = i;
      break;
    }
  if (j != -1)
    return 0; /* already processed */
  if (0 != stat(filename,
		&sbuf)) {
    printf(_("Call to '%s' for file '%s' failed: %s\n"),
	   "stat",
	   filename,
	   strerror(errno));
    return 0;
  }
  if (S_ISREG(sbuf.st_mode)) {
    if (very_verbose)
      printf(_("Processing '%s'\n"),
	     filename);
    return buildIndex(dic->elist,
		      dic->logFile,
		      filename,
		      dic->tree,
		      do_filenames);
  } else
    return 0;
}

static int build(const char * libraries,
		 const char * dbName,
		 size_t mem_limit,
		 const char * log,
		 int argc,
		 char * argv[]) {
  int i;
  int ret;
  DIC cls;
  char * ename;

  if (dbName == NULL) {
    printf(_("No database specified.  Aborting.\n"));
    return -1;
  }
  for (i=strlen(dbName);i>=0;i--) {
    if (dbName[i] == ':') {
      printf(_("'%s' is an invalid database filename (has a colon) for building database (option '%s').\n"),
	     dbName,
	     "-b");
      return -1;
    }
  }
  ename = expandFileName(dbName);
  if (ename == NULL)
    return -1;
  /* unlink(ename); */
  cls.tree = DOODLE_tree_create(&my_log,
				NULL,
				ename);
  free(ename);
  if (cls.tree == NULL)
    return -1;
  if (mem_limit != 0)
    DOODLE_tree_set_memory_limit(cls.tree,
				 mem_limit);

  DOODLE_tree_truncate_modified(cls.tree,
			       &my_log,
			       NULL);

  cls.elist = forkExtractor(do_default,
			    libraries,
			    &my_log,
			    NULL);
  cls.logFile = NULL;
  if (log != NULL) {
    cls.logFile = fopen(log, "w+");
    if (cls.logFile == NULL)
      printf(_("Could not open '%s' for logging: %s.\n"),
	     log,
	     strerror(errno));
  }

  ret = 0;
  for (i=0;i<argc;i++) {
    char * exp;

    if (verbose)
      printf(_("Indexing '%s'\n"),
	     argv[i]);
    exp = expandFileName(argv[i]);
    if (-1 == scanDirectory(exp,
			    &my_log,
			    NULL,
			    &isPruned,
			    NULL,
			    &do_index,
			    &cls)) {
      ret = -1;
      free(exp);
      break;
    }
    free(exp);
  }
  joinExtractor(cls.elist);  
  DOODLE_tree_destroy(cls.tree);
  if (cls.logFile != NULL)
    fclose(cls.logFile);
  return ret;
}

/**
 * @brief closure for printIt
 */
typedef struct {
  struct EXTRACTOR_PluginList * list;
  char ** filenames_seen;
  unsigned int seen_size;
  unsigned int seen_count;
} PrintItArgs;



static void printIt(const DOODLE_FileInfo * fileinfo,
		    PrintItArgs * args) {
  struct EXTRACTOR_PluginList *list = args->list;
  const char * filename;
  int i;

  filename = fileinfo->filename;
  if (isPruned(filename, NULL))
    return;
  for (i=args->seen_count-1;i>=0;i--)
    if (0 == strcmp(filename,
		    args->filenames_seen[i]))
      return;
  if (args->seen_count == args->seen_size) {
    GROW(args->filenames_seen,
	 args->seen_size,
	 args->seen_size*2 + 128);
  }
  args->filenames_seen[args->seen_count++] = STRDUP(filename);
  if (! access(filename, R_OK | F_OK)) {
    if (do_extract) {
      /* print */
      printf(_("Keywords for matching file '%s':\n"),
	     filename);
      EXTRACTOR_extract (list,
			 filename,
			 NULL, 0,
			 &EXTRACTOR_meta_data_print,
			 stdout);
    } else {
      printf("%s\n",
	     filename);
    }
  }
}

static int print(const char * dbName) {
  struct DOODLE_SuffixTree * tree;
  char * ename;
  struct stat buf;
  int ret;

 if (dbName == NULL) {
    printf(_("No database specified. Aborting.\n"));
    return -1;
  }
  ename = expandFileName(dbName);
  if (0 != stat(ename, &buf)) {
    printf(_("Call to '%s' for file '%s' failed: %s.\n"),
	   "stat",
	   dbName,
	   strerror(errno));
    free(ename);
    return -1;
  }
  tree = DOODLE_tree_open_RDONLY(&my_log,
				 NULL,
				 ename);
  free(ename);
  if (tree == NULL)
    return -1;
  ret = DOODLE_tree_dump(stdout,
			 tree);
  DOODLE_tree_destroy(tree);
  return ret;
}

static int search(const char * libraries,
		  const char * dbName,
		  size_t mem_limit,
		  int argc,
		  char * argv[]) {
  int ret;
  struct stat buf;
  char * ename;
  struct DOODLE_SuffixTree * tree;
  struct EXTRACTOR_PluginList * extractors;
  int i;
  PrintItArgs args;
  char * utf;

  if (dbName == NULL) {
    printf(_("No database specified. Aborting.\n"));
    return -1;
  }
  ename = expandFileName(dbName);
  if (0 != stat(ename, &buf)) {
    printf(_("Call to '%s' for file '%s' failed: %s.\n"),
	   "stat",
	   dbName,
	   strerror(errno));
    free(ename);
    return -1;
  }
  tree = DOODLE_tree_open_RDONLY(&my_log,
				 NULL,
				 ename);
  if (mem_limit != 0)
    DOODLE_tree_set_memory_limit(tree,
				 mem_limit);

  free(ename);
  if (tree == NULL)
    return -1;
  if (do_extract) {
    if (do_default)
      extractors = EXTRACTOR_plugin_add_defaults(EXTRACTOR_OPTION_DEFAULT_POLICY);
    else
      extractors = NULL;
    if (libraries != NULL)
      extractors = EXTRACTOR_plugin_add_config(extractors,
					       libraries,
					       EXTRACTOR_OPTION_DEFAULT_POLICY);
  } else
    extractors = NULL;

  ret = 0;
  args.list = extractors;
  args.filenames_seen = NULL;
  args.seen_size = 0;
  args.seen_count = 0;

  for (i=0;i<argc;i++) {
    printf(_("Searching for '%s':\n"),
	   argv[i]);
    if (strlen(argv[i]) > MAX_LENGTH) {
      printf(_("Warning: search string is longer than %d characters, search will not work.\n"),
	     MAX_LENGTH);
      continue; /* no need to even try... */
    }
    if (strlen(argv[i]) > MAX_LENGTH/2) {
      printf(_("Warning: search string is longer than %d characters, search may not work properly.\n"),
	     MAX_LENGTH/2);
    }
    utf = convertToUtf8(argv[i],
			strlen(argv[i]),
			nl_langinfo(CODESET));
    if ( (do_approx == 0) &&
	 (ignore_case == 0) ) {
      if (0 == DOODLE_tree_search(tree,
				  utf,
				  (DOODLE_ResultCallback) &printIt,
				  &args)) {
	printf(_("\tNot found!\n"));
	ret++;
      }
    } else {
      if (0 == DOODLE_tree_search_approx(tree,
					 do_approx,
					 ignore_case,
					 utf,
					 (DOODLE_ResultCallback) &printIt,
					 &args)) {
	printf(_("\tNot found!\n"));
	ret++;
      }
    }
    free(utf);
  }
  DOODLE_tree_destroy(tree);
  EXTRACTOR_plugin_remove_all(extractors);

  for (i=0;i<args.seen_count;i++)
    free(args.filenames_seen[i]);
  GROW(args.filenames_seen,
       args.seen_size,
       0);
  return ret;
}

int main(int argc,
	 char * argv[]) {
  int c;
  int option_index;
  char * libraries = NULL;
  int do_build = 0;
  unsigned int mem_limit = 0;
  char * dbName;
  char * tmp;
  char * log = NULL;
  int ret;

  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  dbName = getenv("DOODLE_PATH");
  if (NULL == dbName)
    dbName = "~/.doodle";
  tmp = getenv("PRUNEPATHS");
  if (tmp != NULL)
    prunepaths = tmp;

  while (1) {
    static struct option long_options[] = {
      {"approximate", 1, 0, 'a'},
      {"build", 0, 0, 'b'},
      {"database", 1, 0, 'd'},
      {"extract", 0, 0, 'e'},
      {"filenames", 0, 0, 'f'} ,
      {"help", 0, 0, 'h'},
      {"ignore-case", 0, 0, 'i'},
      {"library", 1, 0, 'l'},
      {"log", 1, 0, 'L'},
      {"memory", 1, 0, 'm'},
      {"nodefault", 1, 0, 'n'},
      {"prunepaths", 1, 0, 'P' },
      {"print", 0, 0, 'p'},
      {"verbose", 0, 0, 'V'},
      {"version", 0, 0, 'v'},
      {NULL, 0, 0, 0}
    };
    option_index = 0;
    c = getopt_long(argc,
		    argv, "a:bd:efhil:L:m:nP:pVv",
		    long_options,
		    &option_index);

    if (c == -1)
      break; /* No more flags to process */
    switch (c) {
    case 'a':
      if (1 != sscanf(optarg, "%ud", &do_approx)) {
	printf(_("You must pass a number to the '%s' option.\n"),
	       "-a");
	return -1;
      }
      if (do_build == 1) {
	printf(_("The options '%s' and '%s' cannot be used together!\n"),
	       "-a", "-b");
	return -1;
      }	
      break;
    case 'b':
      do_build = 1;
      if (do_approx != 0) {
	printf(_("The options '%s' and '%s' cannot be used together!\n"),
	       "-a", "-b");
	return -1;
      }	
      if (do_print == 1) {
	printf(_("The options '%s' and '%s' cannot be used together!\n"),
	       "-b", "-p");
	return -1;
      }	
      if (ignore_case == 1) {
	printf(_("The options '%s' and '%s' cannot be used together!\n"),
	       "-b", "-i");
	return -1;
      }
      break;
    case 'd':
      dbName = optarg;
      break;
    case 'e':
      do_extract = 1;
      break;
    case 'f':
      do_filenames = 1;
      break;
    case 'h':
      printHelp();
      return 0;
    case 'i':
      ignore_case = 1;
      if (do_build == 1) {
	printf(_("The options '%s' and '%s' cannot be used together!\n"),
	       "-b", "-i");
	return -1;
      }
      break;
    case 'l':
      libraries = optarg;
      break;
    case 'L':
      log = optarg;
      break;
    case 'm':
      if (1 != sscanf(optarg, "%u", &mem_limit)) {
	printf(_("You must pass a number to the '%s' option.\n"),
	       "-m");
	return -1;
      }
      if (mem_limit > 0xFFFFFFFF / 1024 / 1024) {
	printf(_("Specified memory limit is too high.\n"));
	return -1;
      }
      mem_limit *= 1024 * 1024;
      break;
    case 'n':
      do_default = 0;
      break;
    case 'p':
      do_print = 1;
      if (do_build == 1) {
	printf(_("The options '%s' and '%s' cannot be used together!\n"),
	       "-b", "-p");
	return -1;
      }	
      break;
    case 'P':
      prunepaths = optarg;
      break;
    case 'V':
      if (verbose == 1)
	very_verbose = 1;
      verbose = 1;
      break;
    case 'v':
      printf(_("Version %s\n"),
	     PACKAGE_VERSION);
      return 0;
    default:
      fprintf(stderr,
	      _("Use '--help' to get a list of options.\n"));
      return -1;
    }  /* end of parsing commandline */
  } /* while (1) */

  if ( (do_print == 0) &&
       (argc - optind < 1) ) {
    fprintf(stderr,
	    do_build ?
	    _("Invoke with filenames or directories to index!\n") :
	    _("Invoke with query terms to search for!\n"));
    return -1;
  }

  if (libraries != NULL)
    libraries = strdup(libraries);
  if (do_build) {
    ret = build(libraries,
		dbName,
		mem_limit,
		log,
		argc - optind,
		&argv[optind]);
    if (libraries != NULL)
      free(libraries);
    return ret;
  } else if (do_print) {
    int i;
    char * name;

    ret = 0;
    name = strdup(dbName);
    for (i=strlen(name)-1;i>=0;i--) {
      if (name[i] == ':') {
	ret += print(&name[i+1]);		
	name[i] = '\0';
      }
    }
    ret = print(name);
    free(name);
    if (libraries != NULL)
      free(libraries);
    return ret;
  } else {
    int i;
    char * name;

    ret = 0;
    name = strdup(dbName);
    for (i=strlen(name)-1;i>=0;i--) {
      if (name[i] == ':') {
	ret += search(libraries,
		      &name[i+1],
		      mem_limit,
		      argc - optind,
		      &argv[optind]);
	name[i] = '\0';
      }
    }
    ret += search(libraries,
		  name,
		  mem_limit,
		  argc - optind,
		  &argv[optind]);
    free(name);
    if (libraries != NULL)
      free(libraries);
    return ret;
  }
}

/* end of doodle.c */
